### PR TITLE
(#671) send offer letter on approve

### PR DESCRIPTION
**Pull request by abhishek-pokhriyal**
#671 
- To generate PDF, we have used [mPDF](https://github.com/niklasravnsborg/laravel-pdf) package.
- We need to run  `composer update` command to use the package.

### CODE REVIEW

**Comment by abhishek-pokhriyal** [comment]
@rathorevaibhav It looks to me like the requirements mentioned in #671 description are complete. Please have a look.

File: app/Models/HR/ApplicationRound.php [hunk 1]
```diff
@@ -238,6 +266,6 @@ public function isConfirmed()
      */
     public function getShowActionsAttribute()
     {
-        return is_null($this->round_status) || $this->isRejected() || ($this->isConfirmed() && $this->application->isSentForApproval());
+        return is_null($this->round_status) || $this->isRejected() || ($this->isConfirmed() && $this->application->isSentForApproval() || $this->application->isApproved()) ;
```
Comment (rathorevaibhav) [inline app/Models/HR/ApplicationRound.php]: In what case it should not show the actions?

File: resources/assets/js/app.js [hunk 2]
```diff
@@ -23,6 +23,21 @@ Vue.component('project-stage-billing-component', require('./components/ProjectSt
 Vue.component('invoice-project-component', require('./components/InvoiceProjectComponent.vue'));
 Vue.component('applicant-round-action-component', require('./components/HR/ApplicantRoundActionComponent.vue'));
 
+$(document).ready(() => {
```
Comment (rathorevaibhav) [inline resources/assets/js/app.js]: why this change?

File: resources/views/hr/application/approve-applicant-modal.blade.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
+<div class="modal fade" id="approve_applicant" tabindex="-1" role="dialog" aria-labelledby="approve_applicant" aria-hidden="true" v-if="selectedAction == 'approve'">
```
Comment (rathorevaibhav) [inline resources/views/hr/application/approve-applicant-modal.blade.php]: I believe the approve action is for `application` not an applicant. This should be renamed.

File: resources/views/hr/application/edit.blade.php [hunk 4]
```diff
@@ -183,7 +182,11 @@
                                 if ($loop->last) {
                                     if ($applicationRound->application->status == config('constants.hr.status.sent-for-approval.label')) {
                                         $showFooter = true;
-                                    } elseif (in_array($applicationRound->round_status, [null, config('constants.hr.status.rejected.label')])) {
+                                    }
+                                     elseif ($applicationRound->application->status== config('constants.hr.status.approved.label')) {
+                                        $showFooter = true;
+                                    }
+                                     elseif (in_array($applicationRound->round_status, [null, config('constants.hr.status.rejected.label')])) {
```
Comment (rathorevaibhav) [inline resources/views/hr/application/edit.blade.php]: is this PSR2 formatted?

File: resources/views/hr/application/edit.blade.php [hunk 5]
```diff
@@ -183,7 +182,11 @@
                                 if ($loop->last) {
                                     if ($applicationRound->application->status == config('constants.hr.status.sent-for-approval.label')) {
                                         $showFooter = true;
-                                    } elseif (in_array($applicationRound->round_status, [null, config('constants.hr.status.rejected.label')])) {
+                                    }
+                                     elseif ($applicationRound->application->status== config('constants.hr.status.approved.label')) {
```
Comment (rathorevaibhav) [inline resources/views/hr/application/edit.blade.php]: PSR2?

File: resources/views/hr/application/index.blade.php [hunk 6]
```diff
@@ -79,6 +71,15 @@
                 @endif
                 </a>
             </li>
+            <li class="nav-item">
```
Comment (rathorevaibhav) [inline resources/views/hr/application/index.blade.php]: why this change?

File: resources/views/hr/application/timeline.blade.php [hunk 7]
```diff
@@ -25,9 +24,8 @@
                         @if ($applicationRound->mail_sent)
                             <span data-toggle="modal" data-target="#{{ $applicationRound->communicationMail['modal-id'] }}" class="{{ config("constants.hr.status.$applicationRound->round_status.class") }} modal-toggler">Communication mail</span><br>
                             @include('hr.communication-mail-modal', [ 'data' => $applicationRound->communicationMail ])
-                        @endif
+                            @endif
```
Comment (rathorevaibhav) [inline resources/views/hr/application/timeline.blade.php]: remove whitespaces.

File: resources/views/hr/application/index.blade.php (hunk 6 above)
Comment (vikasrinvi) [reply resources/views/hr/application/index.blade.php]: This changes made to changes the sequnce of `Status` to look more coherence . 

## Initially Its looks like this

comment
## Now it looks like this
comment1

File: resources/assets/js/app.js (hunk 2 above)
Comment (abhishek-pokhriyal) [reply resources/assets/js/app.js]: I think should be at the top because this needs to be loaded first to work properly. Without this, the status alert that comes after the application was approved was not working.

File: app/Models/HR/ApplicationRound.php (hunk 1 above)
Comment (abhishek-pokhriyal) [reply app/Models/HR/ApplicationRound.php]: @rathorevaibhav after the applicant is onboarded (not sure). Earlier, it was not showing dropdown to take further actions when the application was approved. So there was no way to onboard the applicant.

File: resources/assets/js/app.js (hunk 2 above)
Comment (abhishek-pokhriyal) [reply resources/assets/js/app.js]: @rathorevaibhav we had a problem like this earlier also. This was when the status alert would not close in the book module's index page.

File: app/Models/HR/ApplicationRound.php (hunk 1 above)
Comment (rathorevaibhav) [reply app/Models/HR/ApplicationRound.php]: I think then it's easy to check for one negative case rather than checking for five positive checks. Thoughts?

File: app/Http/Controllers/HR/Applications/ApplicationController.php [hunk 8]
```diff
@@ -89,6 +96,21 @@ public function edit($id)
         return view('hr.application.edit')->with($attr);
     }
 
+    public static function generateOfferLetter($id, $redirect = true)
```
Comment (abhishek-pokhriyal) [inline app/Http/Controllers/HR/Applications/ApplicationController.php]: @Satendra-SR this function should be somewhere else. thoughts?

File: resources/assets/js/app.js [hunk 9]
```diff
@@ -218,20 +233,6 @@ $('.date-field').datepicker({
     dateFormat: "dd/mm/yy"
 });
 
-$(document).ready(() => {
```
Comment (Satendra-SR) [inline resources/assets/js/app.js]: @abhishek-pokhriyal can you check the necessity of these changes?

**Comment by abhishek-pokhriyal** [comment]
@rathorevaibhav we have completed the development of sent for approval, approve and onboard. Please review the code and provide your feedback.

File: app/User.php [hunk 10]
```diff
@@ -82,4 +82,26 @@ public function scopeFindByEmail($query, $email)
     {
         return $query->where('email', $email);
     }
+
+    public static function _create($attr, $applicant = null)
+    {
+        $name = isset($attr['name']) ? $attr['name'] : null;
+        $email = isset($attr['email']) ? $attr['email'] : $attr['onboard_email'] . '@' . env('GOOGLE_CLIENT_HD');
```
Comment (rathorevaibhav) [inline app/User.php]: email and name should be mandatory. not optional. not nullable.

File: app/User.php [hunk 11]
```diff
@@ -82,4 +82,26 @@ public function scopeFindByEmail($query, $email)
...
+        }
+         
+        if (!$name) {
+            return;
+        }
+
+        $user = self::create([
+            'email' => $email,
+            'name' => $name,
+            'password' => isset($attr['password']) ? $attr['password'] : null,
```
Comment (rathorevaibhav) [inline app/User.php]: either you ask for a password or stores a random one. Null passwords are not allowed.

File: app/User.php [hunk 12]
```diff
@@ -82,4 +82,26 @@ public function scopeFindByEmail($query, $email)
...
+            'provider' => isset($attr['provider']) ? $attr['provider'] : 'google',
```
Comment (rathorevaibhav) [inline app/User.php]: what does this field represent? why storing "google"?

File: app/User.php [hunk 13]
```diff
@@ -82,4 +82,26 @@ public function scopeFindByEmail($query, $email)
...
+            'provider_id' => isset($attr['provider_id']) ? $attr['provider_id'] : null,
```
Comment (rathorevaibhav) [inline app/User.php]: what does this field represent? why storing null?

File: config/app.php [hunk 14]
```diff
@@ -146,6 +146,7 @@
         Illuminate\Translation\TranslationServiceProvider::class,
         Illuminate\Validation\ValidationServiceProvider::class,
         Illuminate\View\ViewServiceProvider::class,
+        niklasravnsborg\LaravelPdf\PdfServiceProvider::class,
```
Comment (rathorevaibhav) [inline config/app.php]: why is this needed? what does adding package's serviceprovider in app.php do?

File: config/app.php [hunk 15]
```diff
@@ -213,6 +214,7 @@
         'Validator' => Illuminate\Support\Facades\Validator::class,
         'View' => Illuminate\Support\Facades\View::class,
         'Socialite' => Laravel\Socialite\Facades\Socialite::class,
+        'PDF' => niklasravnsborg\LaravelPdf\Facades\Pdf::class,
```
Comment (rathorevaibhav) [inline config/app.php]: what does adding an alias in app.php do?

File: database/migrations/2018_08_18_033135_create_offer_letter.php [hunk 16]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+class CreateOfferLetter extends Migration
```
Comment (rathorevaibhav) [inline database/migrations/2018_08_18_033135_create_offer_letter.php]: Ideally the class name should be `CreateOfferLetterColumn`

File: database/migrations/2018_09_01_095015_update_collumn_of_users_table.php [hunk 17]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+use Illuminate\Support\Facades\Schema;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Database\Migrations\Migration;
+
+class UpdateCollumnOfUsersTable extends Migration
```
Comment (rathorevaibhav) [inline database/migrations/2018_09_01_095015_update_collumn_of_users_table.php]: spell check

Comment (rathorevaibhav) [reply database/migrations/2018_09_01_095015_update_collumn_of_users_table.php]: what exactly does it do? Does it update multiple columns? Does it remove a column?

File: database/migrations/2018_09_01_095015_update_collumn_of_users_table.php [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        //
```
Comment (rathorevaibhav) [inline database/migrations/2018_09_01_095015_update_collumn_of_users_table.php]: why is this empty?

File: resources/views/hr/application/offerletter.blade.php [hunk 19]
```diff
@@ -0,0 +1,27 @@
+<!DOCTYPE html>
```
Comment (rathorevaibhav) [inline resources/views/hr/application/offerletter.blade.php]: offer-letter.blade.php

File: resources/views/hr/application/send-for-approval-modal.blade.php [hunk 20]
```diff
@@ -8,9 +8,22 @@
...
+                <div class="form-row">
+                    <div class="form-group col-md-12">
+                        @if ($application->offer_letter)
+                            <div class="col-md-12">
+                                <a target="_blank" href="{{ route("applications.{$application->job->type}.offer-letter", $application) }}" class="d-flex align-items-center">
+                                    <i class="fa fa-file fa-2x text-primary btn-file"></i>&nbsp;See offer letter
+                                </a>
+                            </div>
+                        @else
+                            <a href="{{ route('applications.generateOfferLetter', $application->id) }}" class="btn btn-secondary px-4">Generate Offer Letter</a>
```
Comment (rathorevaibhav) [inline resources/views/hr/application/send-for-approval-modal.blade.php]: are we tracking to fix the UI/UX problem in this later?

File: resources/views/hr/application/timeline.blade.php [hunk 21]
```diff
@@ -49,6 +46,35 @@
...
+                        {{ $event->value->conductedPerson }} requested approval from {{ $event->value->supervisor }}
+                        @break
+                    @case(config('constants.hr.application-meta.keys.approved'))
+                        @php
+                            $event = $item['event'];
+                        @endphp
+                        <b><u>{{ date(config('constants.display_date_format'), strtotime($item['date'])) }}</u></b><br>
+                        {{ $event->value->approvedBy }} approved this application
+                        @break
+                    @case(config('constants.hr.application-meta.keys.approved'))
```
Comment (rathorevaibhav) [inline resources/views/hr/application/timeline.blade.php]: duplicate code?

File: resources/views/settings/hr/approved.blade.php [hunk 22]
```diff
@@ -0,0 +1,31 @@
+<div class="card mt-4">
+	<form action="{{ url('/settings/hr/update') }}" method="POST">
```
Comment (rathorevaibhav) [inline resources/views/settings/hr/approved.blade.php]: does this url have a name?

File: resources/views/settings/hr/approved.blade.php [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+		{{ csrf_field() }}
```
Comment (rathorevaibhav) [inline resources/views/settings/hr/approved.blade.php]: Alternatively, you can use Laravel's blade syntax `@csrf` to have a csrf_field

File: resources/views/settings/hr/approved.blade.php [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+
+		<div class="card-header c-pointer" data-toggle="collapse" data-target="#approved_email_template" aria-expanded="true" aria-controls="approved_email_template">Approve email to Applicant</div>
```
Comment (rathorevaibhav) [inline resources/views/settings/hr/approved.blade.php]: `Approved email to applicant`

File: routes/web.php [hunk 25]
```diff
@@ -62,9 +62,21 @@
...
+
+            Route::get('job/{application}/offer-letter', 'JobApplicationController@viewOfferLetter')->name('applications.job.offer-letter');
+            Route::get('internship/{application}/offer-letter', 'InternshipApplicationController@viewOfferLetter')->name('applications.internship.offer-letter');
+            Route::get('volunteer/{application}/offer-letter', 'VolunteerApplicationController@viewOfferLetter')->name('applications.volunteer.offer-letter');
+
             Route::resource('job', 'JobApplicationController')
                 ->only(['index', 'edit', 'update'])
                 ->names(['index' => 'applications.job.index', 'edit' => 'applications.job.edit', 'update' => 'applications.job.update']);
+
+
```
Comment (rathorevaibhav) [inline routes/web.php]: remove extra lines.

File: routes/web.php [hunk 26]
```diff
@@ -62,9 +62,21 @@
...
+
+            Route::get('{id}/generate-offer-letter', 'JobApplicationController@generateOfferLetter')->name('applications.generateOfferLetter');
```
Comment (rathorevaibhav) [inline routes/web.php]: `Route::get('{application}/generate-offer-letter'`?

File: app/Http/Controllers/HR/Applications/ApplicationController.php (hunk 8 above)
Comment (rathorevaibhav) [inline app/Http/Controllers/HR/Applications/ApplicationController.php]: Should be
```
public static function generateOfferLetter(Application $application, boolean $redirect = true)
```

File: app/Http/Controllers/HR/Applications/ApplicationController.php [hunk 27]
```diff
@@ -89,6 +96,21 @@ public function edit($id)
...
+    {
+        $application = Application::findOrFail($id);
```
Comment (rathorevaibhav) [inline app/Http/Controllers/HR/Applications/ApplicationController.php]: won't need this if you type-hint in the method parameters.

File: app/Http/Controllers/HR/Applications/ApplicationController.php (hunk 8 above)
Comment (rathorevaibhav) [reply app/Http/Controllers/HR/Applications/ApplicationController.php]: have commented about this in my review below.

File: .gitignore [hunk 28]
```diff
@@ -18,4 +18,4 @@ yarn-error.log
 /public/css
 /public/js
 /public/images/vendor
-.DS_Store
```
[the remark this answers is not in the document]
Comment (mohitgusain) [reply .gitignore]: It's by mistake, we will add this again

File: app/Helpers/FileHelper.php [hunk 29]
```diff
@@ -2,8 +2,11 @@
 
 namespace App\Helpers;
 
+use App\Models\HR\Applicant;
 use Carbon\Carbon;
+use Illuminate\Http\UploadedFile;
```
[the remark this answers is not in the document]
Comment (Satendra-SR) [reply app/Helpers/FileHelper.php]: previously, we were using UploadedFile's object to store `uploaded-offer` letter, but now we are using PDF's object, so it should be removed too.

File: app/Http/Controllers/HR/Applications/ApplicationController.php [hunk 30]
```diff
@@ -69,13 +73,16 @@ public function edit($id)
         $application->load(['evaluations', 'evaluations.evaluationParameter', 'evaluations.evaluationOption', 'job', 'job.rounds', 'job.rounds.evaluationParameters', 'job.rounds.evaluationParameters.options', 'applicant', 'applicant.applications', 'applicationRounds', 'applicationRounds.evaluations', 'applicationRounds.round', 'applicationMeta']);
 
         $job = $application->job;
+        $approveMailTemplete = Setting::getApprovedEmail();
```
[the remark this answers is not in the document]
Comment (Satendra-SR) [reply app/Http/Controllers/HR/Applications/ApplicationController.php]: This method is retrieving Email template for Approve Modal. it is a bit confusing so we are changing it as `getApplicationApprovedEmail()`.

File: resources/views/hr/application/offerletter.blade.php (hunk 19 above)
Comment (mohitgusain) [reply resources/views/hr/application/offerletter.blade.php]: changes done

File: database/migrations/2018_09_01_095015_update_collumn_of_users_table.php (hunk 18 above)
Comment (mohitgusain) [reply database/migrations/2018_09_01_095015_update_collumn_of_users_table.php]: No need of this migration anymore.

File: app/Models/HR/ApplicationRound.php [hunk 31]
```diff
@@ -73,24 +76,60 @@ public function _update($attr)
...
+                    ]),
+                ]);
+
                 // The below env call needs to be changed to config after the default
                 // credentials bug in the Google API services is resolved.
                 $email = $attr['onboard_email'] . '@' . env('GOOGLE_CLIENT_HD');
                 $applicant->onboard($email, $attr['onboard_password'], [
                     'designation' => $attr['designation'],
                 ]);
+                $user = User::_create($attr, $applicant);
```
[the remark this answers is not in the document]
Comment (vikasrinvi) [reply app/Models/HR/ApplicationRound.php]: Sure, I have made changes. Now we are using `User::create($params)`  directly in `ApplicationRound.php`

File: app/User.php (hunk 12 above)
Comment (vikasrinvi) [reply app/User.php]: These fields represent that which service does the current user is using. And it is storing `google` because we are using `G-suite` which is provided by `Google`.

File: app/User.php (hunk 13 above)
Comment (vikasrinvi) [reply app/User.php]: `provider_id` is not needed so now after updating, we are sending empty string. later we will remove this `provider_id` in the issue number #677 .

**Comment by abhishek-pokhriyal** [comment]
@rathorevaibhav done with your feedbacks

File: database/migrations/2018_09_01_095015_update_collumn_of_users_table.php (hunk 17 above)
Comment (vikasrinvi) [reply database/migrations/2018_09_01_095015_update_collumn_of_users_table.php]: This migration was made for changing the default value of provider_id but as now it is not required so we have removed this migration.

File: resources/views/hr/application/send-for-approval-modal.blade.php (hunk 20 above)
Comment (vikasrinvi) [reply resources/views/hr/application/send-for-approval-modal.blade.php]: yes, we will track this problem in issue #678 .

File: resources/views/hr/application/timeline.blade.php (hunk 21 above)
Comment (vikasrinvi) [reply resources/views/hr/application/timeline.blade.php]: We have removed the redundant code

File: config/app.php (hunk 14 above)
Comment (Satendra-SR) [reply config/app.php]: In Providers array we can list the class names of the package's service providers. but here, it is not needed, In Package's `Readme` file `package supports auto-discovery features of Laravel 5.5+` was mentioned.

File: app/Helpers/FileHelper.php [hunk 32]
```diff
@@ -36,4 +39,23 @@ public static function getCurrentStorageDirectory()
...
+        return "$dashedApplicantName-$timestamp.pdf";
+    }
+
+    public static function generateOfferLetter(Application $application)
+    {
+        $job = $application->job;
+        $applicant = $application->applicant;
+        $pdf = PDF::loadView('hr.application.offer-letter', compact('applicant', 'job'));
+        $fileName = self::getOfferLetterFileName($pdf, $applicant);
+        $full_path = storage_path('app/' . config('constants.hr.offer-letters-dir') . '/' . $fileName);
```
[the remark this answers is not in the document]
Comment (Satendra-SR) [reply app/Helpers/FileHelper.php]: done

File: app/Helpers/FileHelper.php [hunk 33]
```diff
@@ -36,4 +39,23 @@ public static function getCurrentStorageDirectory()
         $now = Carbon::now();
         return $now->format('Y') . '/' . $now->format('m');
     }
+
+    public static function getOfferLetterFileName($file, Applicant $applicant)
```
[the remark this answers is not in the document]
Comment (Satendra-SR) [reply app/Helpers/FileHelper.php]: done

File: resources/views/hr/application/send-for-approval-modal.blade.php [hunk 34]
```diff
@@ -8,9 +8,22 @@
                 </button>
             </div>
             <div class="modal-body">
+                <div class="form-row">
+                    <div class="form-group col-md-12">
+                        @if ($application->offer_letter)
+                            <div class="col-md-12">
```
[the remark this answers is not in the document]
Comment (mohitgusain) [reply resources/views/hr/application/send-for-approval-modal.blade.php]: Yes, this is redundant.

File: config/app.php [hunk 35]
```diff
@@ -11,7 +11,7 @@
     | framework needs to place the application's name in a notification or
```
[the remark this answers is not in the document]
Comment (vikasrinvi) [reply config/app.php]: yes we have removed this
